### DISCUSSION TITLE
Support for SMB over QUIC and Generic UDP Proxying in YARP

### DISCUSSION THREAD

**Discussion by krasiaa123** [dotnet/yarp#2654] [Ideas] [answered]
YARP currently excels at reverse proxying HTTP-based traffic, including support for HTTP/1.1, HTTP/2, and HTTP/3 (QUIC). However, as protocols like SMB over QUIC (introduced with Windows Server 2022 and Azure SMB support, now Windows Server 2025 supports SMB over QUIC on-premise) gain traction in modern networking, there is an increasing need for generic UDP proxying capabilities within YARP.

SMB over QUIC operates entirely over UDP on port 443, providing secure and fast file sharing without the need for VPNs. While YARP supports QUIC for HTTP/3, it does not yet allow proxying of other UDP-based protocols like SMB over QUIC.

This feature would expand YARP's versatility by enabling on-premises reverse proxying of UDP-based protocols, such as SMB over QUIC, DNS over QUIC and other custom QUIC-based applications

**Comment by Tratcher** [comment]
YARP is fundamentally an HTTP proxy, there are almost no components that would be re-usable with other protocols. E.g. I tried using YARP to write a TCP proxy but ended up only re-using one file, the StreamCopier, with modifications.

Yes, other protocols need proxies too, but there's little advantage to bundling those with YARP.

**Reply by krasiaa123** [reply]
Well, SMB over QUIC leverages HTTP/3 and QUIC, with the key difference being its use of HTTP over UDP instead of HTTP over TCP. It’s essentially a different transport layer implementation while still adhering to HTTP/3 standards.

For instance, both NGINX and HAProxy already support similar functionality, enabling proxying of HTTP over UDP.

Unfortunately, Microsoft’s enterprise Web Application Proxy (on-premise) also falls short in this area, as it does not support HTTP over UDP or HTTP servers with custom HTTP methods. For example, attempting to reverse proxy an SSTP server through Web Application Proxy is currently impossible due to the lack of support for custom HTTP methods.

**Reply by Tratcher** [reply, marked as answer]
Ah, bad assumption on my part. HTTP/3 is always over QUIC/UDP, no? What's preventing this from working today then?
